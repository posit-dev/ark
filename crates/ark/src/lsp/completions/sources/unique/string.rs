//
// string.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

use anyhow::Result;
use tower_lsp::lsp_types::CompletionItem;

use super::file_path::completions_from_file_path;
use crate::lsp::document_context::DocumentContext;
use crate::treesitter::NodeTypeExt;

pub fn completions_from_string(context: &DocumentContext) -> Result<Option<Vec<CompletionItem>>> {
    log::info!("completions_from_string()");

    let node = context.node;

    if !node.is_string() {
        return Ok(None);
    }

    // Must actually be "inside" the string, so these places don't count, even
    // though they are detected as part of the string nodes `|""|`
    if node.start_position() == context.point || node.end_position() == context.point {
        return Ok(None);
    }

    // Even if we don't find any completions, we were inside a string so we
    // don't want to provide completions for anything else, so we always at
    // least return an empty `completions` vector from here.
    let mut completions: Vec<CompletionItem> = vec![];

    // Return empty set if we are here due to a trigger character like `$`.
    // See posit-dev/positron#1884.
    if context.trigger.is_some() {
        return Ok(Some(completions));
    }

    // Try file path completions
    completions.append(&mut completions_from_file_path(context)?);

    Ok(Some(completions))
}

#[cfg(test)]
mod tests {
    use harp::assert_match;
    use tree_sitter::Point;

    use crate::lsp::completions::sources::completions_from_unique_sources;
    use crate::lsp::completions::sources::unique::string::completions_from_string;
    use crate::lsp::document_context::DocumentContext;
    use crate::lsp::documents::Document;
    use crate::test::r_test;
    use crate::treesitter::NodeTypeExt;

    #[test]
    fn test_outside_quotes() {
        r_test(|| {
            // Before or after the `''`, i.e. `|''` or `''|`.
            // Still considered part of the string node.
            let point = Point { row: 0, column: 0 };
            let document = Document::new("''", None);
            let context = DocumentContext::new(&document, point, None);

            assert!(context.node.is_string());
            assert_eq!(completions_from_string(&context).unwrap(), None);
        })
    }

    #[test]
    fn test_not_string() {
        r_test(|| {
            let point = Point { row: 0, column: 0 };
            let document = Document::new("foo", None);
            let context = DocumentContext::new(&document, point, None);

            assert!(context.node.is_identifier());
            assert_eq!(completions_from_string(&context).unwrap(), None);
        })
    }

    #[test]
    fn test_trigger() {
        r_test(|| {
            // Before or after the `''`, i.e. `|''` or `''|`.
            // Still considered part of the string node.
            let point = Point { row: 0, column: 2 };

            // Assume home directory is not empty
            let document = Document::new("'~/'", None);

            // `None` trigger -> Return file completions
            let context = DocumentContext::new(&document, point, None);
            assert_match!(
                completions_from_string(&context).unwrap(),
                Some(items) => {
                    assert!(items.len() > 0)
                }
            );

            // `Some` trigger -> Should return empty completion set
            let context = DocumentContext::new(&document, point, Some(String::from("$")));
            let res = completions_from_string(&context).unwrap();
            assert_match!(res, Some(items) => { assert!(items.len() == 0) });

            // Check one level up too
            let res = completions_from_unique_sources(&context).unwrap();
            assert_match!(res, Some(items) => { assert!(items.len() == 0) });
        })
    }
}
