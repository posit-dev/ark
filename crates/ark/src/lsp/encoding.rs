//
// encoding.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

use ropey::Rope;
use tower_lsp::lsp_types::Position;
use tree_sitter::Point;

/// `PositionEncodingKind` describes the encoding used for the `Position` `character`
/// column offset field. The `Position` `line` field is encoding agnostic, but the
/// `character` field specifies the number of characters offset from the beginning of
/// the line, and the "character" size is dependent on the encoding. The LSP specification
/// states:
///
/// - UTF8: Character offsets count UTF-8 code units (e.g. bytes).
/// - UTF16: Character offsets count UTF-16 code units (default).
/// - UTF32: Character offsets count UTF-32 code units (these are the same as Unicode
///   codepoints, so this `PositionEncodingKind` may also be used for an encoding-agnostic
///   representation of character offsets.)
///
/// The `vscode-languageclient` library that Positron uses on the frontend to create the
/// `Client` side of the LSP currently ONLY supports `UTF16`, and will error on anything
/// else. Their reasoning is that it is easier for the server (ark) to do the re-encoding,
/// since we are tracking the full document state. Track support for UTF-8 here:
/// https://github.com/microsoft/vscode-languageserver-node/issues/1224
///
/// The other interesting part of this is that `TextDocumentContentChangeEvent`s that
/// come through the `did_change()` event and the `TextDocumentItem` that comes through
/// the `did_open()` event encode the `text` of the change/document in UTF-8, even though
/// the `Range` (in the case of `did_change()`) that tells you where to apply the change
/// uses UTF-16, so that's cool. UTF-8 `text` is forced to come through due to how the
/// LSP specification uses jsonrpc, where the content fields must be 'utf-8' encoded:
/// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#contentPart
/// This at least means we have a guarantee that the document itself and any updates to
/// it will be encoded in UTF-8, even if the Positions are UTF-16.
///
/// So we need a way to convert the UTF-16 `Position`s to UTF-8 `tree_sitter::Point`s and
/// back. This requires the document itself, and is what the helpers in this file implement.
pub fn get_position_encoding_kind() -> tower_lsp::lsp_types::PositionEncodingKind {
    tower_lsp::lsp_types::PositionEncodingKind::UTF16
}

pub fn convert_tree_sitter_range_to_lsp_range(
    x: &Rope,
    range: tree_sitter::Range,
) -> tower_lsp::lsp_types::Range {
    let start = convert_point_to_position(x, range.start_point);
    let end = convert_point_to_position(x, range.end_point);
    tower_lsp::lsp_types::Range::new(start, end)
}

pub fn convert_position_to_point(x: &Rope, position: Position) -> Point {
    let line = position.line as usize;
    let character = position.character as usize;

    let character = with_line(x, line, character, convert_character_from_utf16_to_utf8);

    Point::new(line, character)
}

pub fn convert_point_to_position(x: &Rope, point: Point) -> Position {
    let line = point.row;
    let character = point.column;

    let character = with_line(x, line, character, convert_character_from_utf8_to_utf16);

    let line = line as u32;
    let character = character as u32;

    Position::new(line, character)
}

fn with_line<F>(x: &Rope, line: usize, character: usize, f: F) -> usize
where
    F: FnOnce(&str, usize) -> usize,
{
    let Some(x) = x.get_line(line) else {
        let n = x.len_lines();
        let x = x.to_string();
        let line = line + 1;
        // Forcing a full capture so we can learn the situations in which this occurs
        let trace = std::backtrace::Backtrace::force_capture();
        log::error!(
            "Requesting line {line} but only {n} lines exist.\n\nDocument:\n{x}\n\nBacktrace:\n{trace}"
        );
        return 0;
    };

    // If the line is fully contained in a single chunk (likely is), use free conversion to `&str`
    if let Some(x) = x.as_str() {
        return f(x, character);
    }

    // Otherwise, use ever so slightly more expensive String materialization of the
    // line spread across chunks
    let x = x.to_string();
    let x = x.as_str();

    f(x, character)
}

/// Converts a character offset into a particular line from UTF-16 to UTF-8
fn convert_character_from_utf16_to_utf8(x: &str, character: usize) -> usize {
    if x.is_ascii() {
        // Fast pass
        return character;
    }

    // Initial check, since loop would skip this case
    if character == 0 {
        return character;
    }

    let mut n = 0;

    // For each `u32` sized `char`, figure out the equivalent size in UTF-16
    // world of that `char`. Once we hit the requested number of `character`s,
    // that means we have indexed into `x` to the correct position, at which
    // point we can take the current bytes based `pos` that marks the start of
    // this `char`, and add on its UTF-8 based size to return an adjusted column
    // offset. We use `==` because I'm fairly certain they should always align
    // exactly, and it would be good to log if that isn't the case.
    for (pos, char) in x.char_indices() {
        n += char.len_utf16();

        if n == character {
            return pos + char.len_utf8();
        }
    }

    log::error!("Failed to locate UTF-16 offset of {character}. Line: '{x}'.");
    return 0;
}

/// Converts a character offset into a particular line from UTF-8 to UTF-16
fn convert_character_from_utf8_to_utf16(x: &str, character: usize) -> usize {
    if x.is_ascii() {
        // Fast pass
        return character;
    }

    // The UTF-8 -> UTF-16 case is slightly simpler. We just slice into `x`
    // using our existing UTF-8 offset, reencode the slice as a UTF-16 based
    // iterator, and count up the pieces.
    match x.get(..character) {
        Some(x) => x.encode_utf16().count(),
        None => {
            let n = x.len();
            log::error!(
                "Tried to take UTF-8 character {character}, but only {n} characters exist. Line: '{x}'."
            );
            0
        },
    }
}
