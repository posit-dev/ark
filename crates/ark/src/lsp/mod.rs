//
// mod.rs
//
// Copyright (C) 2022 Posit Software, PBC. All rights reserved.
//
//

pub mod backend;
pub mod browser;
pub mod comm;
pub mod completions;
pub mod definitions;
pub mod diagnostics;
pub mod documents;
pub mod editor;
pub mod errors;
pub mod events;
pub mod globals;
pub mod handler;
pub mod help;
pub mod help_proxy;
pub mod hover;
pub mod indexer;
pub mod markdown;
pub mod modules;
pub mod references;
pub mod show_message;
pub mod signature_help;
pub mod symbols;
pub mod traits;
pub mod util;
