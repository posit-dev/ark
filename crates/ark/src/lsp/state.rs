use std::sync::Arc;

use dashmap::DashMap;
use parking_lot::Mutex;
use url::Url;

use crate::lsp::documents::Document;

// This is a work in progress. Currently contains synchronised objects but this
// shouldn't be necessary after we have implemented synchronisation of LSP handlers.
// The handlers will either get a shared ref or an exclusive ref to the worldstate.

#[derive(Clone, Default, Debug)]
/// The world state, i.e. all the inputs necessary for analysing or refactoring code.
pub struct WorldState {
    /// Watched documents
    pub documents: Arc<DashMap<Url, Document>>,

    /// Watched folders
    pub workspace: Arc<Mutex<Workspace>>,

    /// The scopes for the console. This currently contains a list (outer `Vec`)
    /// of names (inner `Vec`) of the environments on the search path, starting
    /// from the global environment and ending with the base package. Eventually
    /// this might also be populated with the scope for the current environment
    /// in debug sessions (not implemented yet).
    ///
    /// This is currently one of the main sources of known symbols for
    /// diagnostics. In the future we should better delineate interactive
    /// contexts (e.g. the console, but scripts might also be treated as
    /// interactive, which could be a user setting) and non-interactive ones
    /// (e.g. a package). In non-interactive contexts, the lexical scopes
    /// examined for diagnostics should be fully determined by variable bindings
    /// and imports (code-first diagnostics).
    ///
    /// In the future this should probably become more complex with a list of
    /// either symbol names (as is now the case) or named environments, such as
    /// `pkg:ggplot2`. Storing named environments here will allow the LSP to
    /// retrieve the symbols in a pull fashion (the whole console scopes are
    /// currently pushed to the LSP), and cache the symbols with Salsa. The
    /// performance is not currently an issue but this could change once we do
    /// more analysis of symbols in the search path.
    pub console_scopes: Arc<Mutex<Vec<Vec<String>>>>,

    /// Currently installed packages
    pub installed_packages: Arc<Mutex<Vec<String>>>,
}

#[derive(Default, Debug)]
pub struct Workspace {
    pub folders: Vec<Url>,
}
