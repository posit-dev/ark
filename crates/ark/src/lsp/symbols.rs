//
// symbols.rs
//
// Copyright (C) 2022-2024 Posit Software, PBC. All rights reserved.
//
//

#![allow(deprecated)]

use std::result::Result::Ok;

use ropey::Rope;
use stdext::unwrap::IntoResult;
use tower_lsp::lsp_types::DocumentSymbol;
use tower_lsp::lsp_types::DocumentSymbolParams;
use tower_lsp::lsp_types::Location;
use tower_lsp::lsp_types::Range;
use tower_lsp::lsp_types::SymbolInformation;
use tower_lsp::lsp_types::SymbolKind;
use tower_lsp::lsp_types::Url;
use tower_lsp::lsp_types::WorkspaceSymbolParams;
use tree_sitter::Node;

use crate::lsp::encoding::convert_point_to_position;
use crate::lsp::indexer;
use crate::lsp::indexer::IndexEntryData;
use crate::lsp::state::WorldState;
use crate::lsp::traits::rope::RopeExt;
use crate::lsp::traits::string::StringExt;
use crate::treesitter::point_end_of_previous_row;
use crate::treesitter::BinaryOperatorType;
use crate::treesitter::NodeType;
use crate::treesitter::NodeTypeExt;

fn new_symbol(name: String, kind: SymbolKind, range: Range) -> DocumentSymbol {
    DocumentSymbol {
        name,
        kind,
        detail: None,
        children: Some(Vec::new()),
        deprecated: None,
        tags: None,
        range,
        selection_range: range,
    }
}

fn new_symbol_node(
    name: String,
    kind: SymbolKind,
    range: Range,
    children: Vec<DocumentSymbol>,
) -> DocumentSymbol {
    let mut symbol = new_symbol(name, kind, range);
    symbol.children = Some(children);
    symbol
}

pub fn symbols(params: &WorkspaceSymbolParams) -> anyhow::Result<Vec<SymbolInformation>> {
    let query = &params.query;
    let mut info: Vec<SymbolInformation> = Vec::new();

    indexer::map(|path, symbol, entry| {
        if !symbol.fuzzy_matches(query) {
            return;
        }

        match &entry.data {
            IndexEntryData::Function { name, arguments: _ } => {
                info.push(SymbolInformation {
                    name: name.to_string(),
                    kind: SymbolKind::FUNCTION,
                    location: Location {
                        uri: Url::from_file_path(path).unwrap(),
                        range: entry.range,
                    },
                    tags: None,
                    deprecated: None,
                    container_name: None,
                });
            },

            IndexEntryData::Section { level: _, title } => {
                info.push(SymbolInformation {
                    name: title.to_string(),
                    kind: SymbolKind::STRING,
                    location: Location {
                        uri: Url::from_file_path(path).unwrap(),
                        range: entry.range,
                    },
                    tags: None,
                    deprecated: None,
                    container_name: None,
                });
            },
            IndexEntryData::Variable { name } => {
                info.push(SymbolInformation {
                    name: name.clone(),
                    kind: SymbolKind::VARIABLE,
                    location: Location {
                        uri: Url::from_file_path(path).unwrap(),
                        range: entry.range,
                    },
                    tags: None,
                    deprecated: None,
                    container_name: None,
                });
            },
        };
    });

    Ok(info)
}

/// Represents a section in the document with its title, level, range, and children
#[derive(Debug)]
struct Section {
    title: String,
    level: usize,
    start_position: tree_sitter::Point,
    end_position: Option<tree_sitter::Point>,
    children: Vec<DocumentSymbol>,
}

pub(crate) fn document_symbols(
    state: &WorldState,
    params: &DocumentSymbolParams,
) -> anyhow::Result<Vec<DocumentSymbol>> {
    let uri = &params.text_document.uri;
    let document = state.documents.get(uri).into_result()?;
    let ast = &document.ast;
    let contents = &document.contents;

    // Start walking from the root node
    let root_node = ast.root_node();
    let mut result = Vec::new();

    // Extract and process all symbols from the AST
    if let Err(err) = collect_symbols(&root_node, contents, 0, &mut result) {
        log::error!("Failed to collect symbols: {err:?}");
        return Ok(Vec::new());
    }

    Ok(result)
}

/// Collect all document symbols from a node recursively
fn collect_symbols(
    node: &Node,
    contents: &Rope,
    current_level: usize,
    symbols: &mut Vec<DocumentSymbol>,
) -> anyhow::Result<()> {
    match node.node_type() {
        NodeType::Program | NodeType::BracedExpression => {
            collect_sections(node, contents, current_level, symbols)?;
        },

        NodeType::Call => {
            collect_call(node, contents, symbols)?;
        },

        NodeType::BinaryOperator(BinaryOperatorType::LeftAssignment) |
        NodeType::BinaryOperator(BinaryOperatorType::EqualsAssignment) => {
            collect_assignment(node, contents, symbols)?;
        },

        // For all other node types, no symbols need to be added
        _ => {},
    }

    Ok(())
}

fn collect_sections(
    node: &Node,
    contents: &Rope,
    current_level: usize,
    symbols: &mut Vec<DocumentSymbol>,
) -> anyhow::Result<()> {
    // In lists of expressions we track and collect section comments, then
    // collect symbols from children nodes

    let mut cursor = node.walk();

    // Track active sections at each level
    let mut active_sections: Vec<Section> = Vec::new();

    for child in node.children(&mut cursor) {
        if let NodeType::Comment = child.node_type() {
            let comment_text = contents.node_slice(&child)?.to_string();

            // If we have a section comment, add it to our stack and close any sections if needed
            if let Some((level, title)) = parse_comment_as_section(&comment_text) {
                let absolute_level = current_level + level;

                // Close any sections with equal or higher level
                while !active_sections.is_empty() &&
                    active_sections.last().unwrap().level >= absolute_level
                {
                    // Set end position for the section being closed
                    if let Some(section) = active_sections.last_mut() {
                        let pos = point_end_of_previous_row(child.start_position(), contents);
                        section.end_position = Some(pos);
                    }
                    finalize_section(&mut active_sections, symbols, contents)?;
                }

                let section = Section {
                    title,
                    level: absolute_level,
                    start_position: child.start_position(),
                    end_position: None,
                    children: Vec::new(),
                };
                active_sections.push(section);
            }

            continue;
        }

        // If we get to this point, `child` is not a section comment.
        // Recurse into child.

        if active_sections.is_empty() {
            // If no active section, extend current vector of symbols
            collect_symbols(&child, contents, current_level, symbols)?;
        } else {
            // Otherwise create new store of symbols for the current section
            let mut child_symbols = Vec::new();
            collect_symbols(&child, contents, current_level, &mut child_symbols)?;

            // Nest them inside last section
            if !child_symbols.is_empty() {
                active_sections
                    .last_mut()
                    .unwrap()
                    .children
                    .extend(child_symbols);
            }
        }
    }

    // Close any remaining active sections
    while !active_sections.is_empty() {
        // Set end position to the parent node's end for remaining sections
        if let Some(section) = active_sections.last_mut() {
            let mut pos = node.end_position();
            if pos.row > section.start_position.row {
                pos = point_end_of_previous_row(pos, contents);
            }
            section.end_position = Some(pos);
        }
        finalize_section(&mut active_sections, symbols, contents)?;
    }

    Ok(())
}

fn collect_call(
    node: &Node,
    contents: &Rope,
    symbols: &mut Vec<DocumentSymbol>,
) -> anyhow::Result<()> {
    let Some(callee) = node.child_by_field_name("function") else {
        return Ok(());
    };
    if !callee.is_identifier() {
        return Ok(());
    }

    let fun_symbol = contents.node_slice(&callee)?.to_string();

    match fun_symbol.as_str() {
        "test_that" => collect_call_test_that(node, contents, symbols)?,
        _ => {},
    }

    Ok(())
}

// https://github.com/posit-dev/positron/issues/1428
fn collect_call_test_that(
    node: &Node,
    contents: &Rope,
    symbols: &mut Vec<DocumentSymbol>,
) -> anyhow::Result<()> {
    let Some(arguments) = node.child_by_field_name("arguments") else {
        return Ok(());
    };

    // We don't do any argument matching and just consider the first argument if
    // a string. First skip over `(`.
    let Some(first_argument) = arguments.child(1).and_then(|n| n.child(0)) else {
        return Ok(());
    };
    if !first_argument.is_string() {
        return Ok(());
    }

    let Some(string) = first_argument.child_by_field_name("content") else {
        return Ok(());
    };

    // Recurse in arguments. We could skip the first one if we wanted.
    let mut children = Vec::new();
    let mut cursor = arguments.walk();
    for child in arguments.children_by_field_name("argument", &mut cursor) {
        if let Some(value) = child.child_by_field_name("value") {
            collect_symbols(&value, contents, 0, &mut children)?;
        }
    }

    let name = contents.node_slice(&string)?.to_string();
    let name = format!("Test: {name}");

    let start = convert_point_to_position(contents, node.start_position());
    let end = convert_point_to_position(contents, node.end_position());

    let symbol = new_symbol_node(name, SymbolKind::FUNCTION, Range { start, end }, children);
    symbols.push(symbol);

    Ok(())
}

fn collect_assignment(
    node: &Node,
    contents: &Rope,
    symbols: &mut Vec<DocumentSymbol>,
) -> anyhow::Result<()> {
    // Check for assignment
    matches!(
        node.node_type(),
        NodeType::BinaryOperator(BinaryOperatorType::LeftAssignment) |
            NodeType::BinaryOperator(BinaryOperatorType::EqualsAssignment)
    )
    .into_result()?;

    // check for lhs, rhs
    let lhs = node.child_by_field_name("lhs").into_result()?;
    let rhs = node.child_by_field_name("rhs").into_result()?;

    // check for identifier on lhs, function on rhs
    let function = lhs.is_identifier_or_string() && rhs.is_function_definition();

    if function {
        return collect_assignment_with_function(node, contents, symbols);
    }

    // otherwise, just index as generic object
    let name = contents.node_slice(&lhs)?.to_string();

    let start = convert_point_to_position(contents, lhs.start_position());
    let end = convert_point_to_position(contents, lhs.end_position());

    let symbol = new_symbol(name, SymbolKind::VARIABLE, Range { start, end });
    symbols.push(symbol);

    Ok(())
}

fn collect_assignment_with_function(
    node: &Node,
    contents: &Rope,
    symbols: &mut Vec<DocumentSymbol>,
) -> anyhow::Result<()> {
    // check for lhs, rhs
    let lhs = node.child_by_field_name("lhs").into_result()?;
    let rhs = node.child_by_field_name("rhs").into_result()?;

    // start extracting the argument names
    let mut arguments: Vec<String> = Vec::new();
    let parameters = rhs.child_by_field_name("parameters").into_result()?;

    let mut cursor = parameters.walk();
    for parameter in parameters.children_by_field_name("parameter", &mut cursor) {
        let name = parameter.child_by_field_name("name").into_result()?;
        let name = contents.node_slice(&name)?.to_string();
        arguments.push(name);
    }

    let name = contents.node_slice(&lhs)?.to_string();
    let detail = format!("function({})", arguments.join(", "));

    let range = Range {
        start: convert_point_to_position(contents, lhs.start_position()),
        end: convert_point_to_position(contents, rhs.end_position()),
    };

    let body = rhs.child_by_field_name("body").into_result()?;

    // Process the function body to extract child symbols
    let mut children = Vec::new();
    collect_symbols(&body, contents, 0, &mut children)?;

    let mut symbol = new_symbol_node(name, SymbolKind::FUNCTION, range, children);
    symbol.detail = Some(detail);
    symbols.push(symbol);

    Ok(())
}

/// Finalize a section by creating a symbol and adding it to the parent section or output
fn finalize_section(
    active_sections: &mut Vec<Section>,
    symbols: &mut Vec<DocumentSymbol>,
    contents: &Rope,
) -> anyhow::Result<()> {
    if let Some(section) = active_sections.pop() {
        let start_pos = section.start_position;
        let end_pos = section.end_position.unwrap_or(section.start_position);

        let range = Range {
            start: convert_point_to_position(contents, start_pos),
            end: convert_point_to_position(contents, end_pos),
        };

        let symbol = new_symbol(section.title, SymbolKind::STRING, range);

        let mut final_symbol = symbol;
        final_symbol.children = Some(section.children);

        if let Some(parent) = active_sections.last_mut() {
            parent.children.push(final_symbol);
        } else {
            symbols.push(final_symbol);
        }
    }

    Ok(())
}

// Function to parse a comment and return the section level and title
pub(crate) fn parse_comment_as_section(comment: &str) -> Option<(usize, String)> {
    // Match lines starting with one or more '#' followed by some non-empty content and must end with 4 or more '-', '#', or `=`
    // Ensure that there's actual content between the start and the trailing symbols.
    if let Some(caps) = indexer::RE_COMMENT_SECTION.captures(comment) {
        let hashes = caps.get(1)?.as_str().len(); // Count the number of '#'
        let title = caps.get(2)?.as_str().trim().to_string(); // Extract the title text without trailing punctuations
        if title.is_empty() {
            return None; // Return None for lines with only hashtags
        }
        return Some((hashes, title)); // Return the level based on the number of '#' and the title
    }

    None
}

#[cfg(test)]
mod tests {
    use tower_lsp::lsp_types::Position;

    use super::*;
    use crate::lsp::documents::Document;

    fn test_symbol(code: &str) -> Vec<DocumentSymbol> {
        let doc = Document::new(code, None);
        let node = doc.ast.root_node();

        let mut symbols = Vec::new();
        collect_symbols(&node, &doc.contents, 0, &mut symbols).unwrap();
        symbols
    }

    #[test]
    fn test_symbol_parse_comment_as_section() {
        assert_eq!(parse_comment_as_section("# foo"), None);
        assert_eq!(parse_comment_as_section("# foo ---"), None);
        assert_eq!(parse_comment_as_section("########"), None);
        assert_eq!(
            parse_comment_as_section("# foo ----"),
            Some((1, String::from("foo")))
        );
    }

    #[test]
    fn test_symbol_comment_sections() {
        assert_eq!(test_symbol("# foo"), vec![]);
        assert_eq!(test_symbol("# foo ---"), vec![]);
        assert_eq!(test_symbol("########"), vec![]);

        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 10,
            },
        };
        assert_eq!(test_symbol("# foo ----"), vec![new_symbol(
            String::from("foo"),
            SymbolKind::STRING,
            range
        )]);
    }

    #[test]
    fn test_symbol_assignment() {
        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 3,
            },
        };
        assert_eq!(test_symbol("foo <- 1"), vec![new_symbol(
            String::from("foo"),
            SymbolKind::VARIABLE,
            range,
        )]);
    }

    #[test]
    fn test_symbol_assignment_function() {
        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 20,
            },
        };

        let mut foo = new_symbol(String::from("foo"), SymbolKind::FUNCTION, range);
        foo.detail = Some(String::from("function()"));

        assert_eq!(test_symbol("foo <- function() {}"), vec![foo]);
    }

    #[test]
    fn test_symbol_assignment_function_nested() {
        let range = Range {
            start: Position {
                line: 0,
                character: 20,
            },
            end: Position {
                line: 0,
                character: 23,
            },
        };
        let bar = new_symbol(String::from("bar"), SymbolKind::VARIABLE, range);

        let range = Range {
            start: Position {
                line: 0,
                character: 0,
            },
            end: Position {
                line: 0,
                character: 30,
            },
        };
        let mut foo = new_symbol(String::from("foo"), SymbolKind::FUNCTION, range);
        foo.children = Some(vec![bar]);
        foo.detail = Some(String::from("function()"));

        assert_eq!(test_symbol("foo <- function() { bar <- 1 }"), vec![foo]);
    }

    #[test]
    fn test_symbol_assignment_function_nested_section() {
        insta::assert_debug_snapshot!(test_symbol(
            "
## title0 ----
foo <- function() {
  # title1 ----
  ### title2 ----
  ## title3 ----
  # title4 ----
}
# title5 ----"
        ));
    }

    #[test]
    fn test_symbol_braced_list() {
        let range = Range {
            start: Position {
                line: 0,
                character: 2,
            },
            end: Position {
                line: 0,
                character: 5,
            },
        };
        let foo = new_symbol(String::from("foo"), SymbolKind::VARIABLE, range);

        assert_eq!(test_symbol("{ foo <- 1 }"), vec![foo]);
    }

    #[test]
    fn test_symbol_section_ranges_extend() {
        let symbols = test_symbol(
            "# Section 1 ----
x <- 1
y <- 2
# Section 2 ----
z <- 3",
        );

        assert_eq!(symbols.len(), 2);

        // First section should extend from line 0 to line 2 (start of next section)
        let section1 = &symbols[0];
        assert_eq!(section1.name, "Section 1");
        assert_eq!(section1.range.start.line, 0);
        assert_eq!(section1.range.end.line, 2);

        // Second section should extend from line 3 to end of file
        let section2 = &symbols[1];
        assert_eq!(section2.name, "Section 2");
        assert_eq!(section2.range.start.line, 3);
        assert_eq!(section2.range.end.line, 3);
    }

    #[test]
    fn test_symbol_section_ranges_in_function() {
        let symbols = test_symbol(
            "foo <- function() {
  # Section A ----
  x <- 1
  y <- 2
  # Section B ----
  z <- 3
}",
        );

        assert_eq!(symbols.len(), 1);

        // Should have one function symbol
        let function = &symbols[0];
        assert_eq!(function.name, "foo");
        assert_eq!(function.kind, SymbolKind::FUNCTION);

        // Function should have two section children
        let children = function.children.as_ref().unwrap();
        assert_eq!(children.len(), 2);

        // First section should extend from line 1 to line 3 (start of next section)
        let section_a = &children[0];
        assert_eq!(section_a.name, "Section A");
        assert_eq!(section_a.range.start.line, 1);
        assert_eq!(section_a.range.end.line, 3);

        // Second section should extend from line 4 to end of function body
        let section_b = &children[1];
        assert_eq!(section_b.name, "Section B");
        assert_eq!(section_b.range.start.line, 4);
        assert_eq!(section_b.range.end.line, 5); // End of function body
    }

    #[test]
    fn test_symbol_call_test_that() {
        insta::assert_debug_snapshot!(test_symbol(
            "
test_that_not('foo', {
  1
})

# title ----

test_that('foo', {
  # title1 ----
  1
  # title2 ----
  foo <- function() {
    2
  }
})

# title2 ----
test_that('bar', {
  1
})
"
        ));
    }
}
