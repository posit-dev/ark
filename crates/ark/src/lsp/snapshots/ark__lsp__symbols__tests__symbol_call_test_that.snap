---
source: crates/ark/src/lsp/symbols.rs
expression: "test_symbol(\"\ntest_that_not('foo', {\n  1\n})\n\n# title ----\n\ntest_that('foo', {\n  # title1 ----\n  1\n  # title2 ----\n  foo <- function() {\n    2\n  }\n})\n\n# title2 ----\ntest_that('bar', {\n  1\n})\n\")"
---
[
    DocumentSymbol {
        name: "title",
        detail: None,
        kind: String,
        tags: None,
        deprecated: None,
        range: Range {
            start: Position {
                line: 5,
                character: 0,
            },
            end: Position {
                line: 15,
                character: 0,
            },
        },
        selection_range: Range {
            start: Position {
                line: 5,
                character: 0,
            },
            end: Position {
                line: 15,
                character: 0,
            },
        },
        children: Some(
            [
                DocumentSymbol {
                    name: "Test: foo",
                    detail: None,
                    kind: Function,
                    tags: None,
                    deprecated: None,
                    range: Range {
                        start: Position {
                            line: 7,
                            character: 0,
                        },
                        end: Position {
                            line: 14,
                            character: 2,
                        },
                    },
                    selection_range: Range {
                        start: Position {
                            line: 7,
                            character: 0,
                        },
                        end: Position {
                            line: 14,
                            character: 2,
                        },
                    },
                    children: Some(
                        [
                            DocumentSymbol {
                                name: "title1",
                                detail: None,
                                kind: String,
                                tags: None,
                                deprecated: None,
                                range: Range {
                                    start: Position {
                                        line: 8,
                                        character: 2,
                                    },
                                    end: Position {
                                        line: 9,
                                        character: 3,
                                    },
                                },
                                selection_range: Range {
                                    start: Position {
                                        line: 8,
                                        character: 2,
                                    },
                                    end: Position {
                                        line: 9,
                                        character: 3,
                                    },
                                },
                                children: Some(
                                    [],
                                ),
                            },
                            DocumentSymbol {
                                name: "title2",
                                detail: None,
                                kind: String,
                                tags: None,
                                deprecated: None,
                                range: Range {
                                    start: Position {
                                        line: 10,
                                        character: 2,
                                    },
                                    end: Position {
                                        line: 13,
                                        character: 3,
                                    },
                                },
                                selection_range: Range {
                                    start: Position {
                                        line: 10,
                                        character: 2,
                                    },
                                    end: Position {
                                        line: 13,
                                        character: 3,
                                    },
                                },
                                children: Some(
                                    [
                                        DocumentSymbol {
                                            name: "foo",
                                            detail: Some(
                                                "function()",
                                            ),
                                            kind: Function,
                                            tags: None,
                                            deprecated: None,
                                            range: Range {
                                                start: Position {
                                                    line: 11,
                                                    character: 2,
                                                },
                                                end: Position {
                                                    line: 13,
                                                    character: 3,
                                                },
                                            },
                                            selection_range: Range {
                                                start: Position {
                                                    line: 11,
                                                    character: 2,
                                                },
                                                end: Position {
                                                    line: 13,
                                                    character: 3,
                                                },
                                            },
                                            children: Some(
                                                [],
                                            ),
                                        },
                                    ],
                                ),
                            },
                        ],
                    ),
                },
            ],
        ),
    },
    DocumentSymbol {
        name: "title2",
        detail: None,
        kind: String,
        tags: None,
        deprecated: None,
        range: Range {
            start: Position {
                line: 16,
                character: 0,
            },
            end: Position {
                line: 19,
                character: 2,
            },
        },
        selection_range: Range {
            start: Position {
                line: 16,
                character: 0,
            },
            end: Position {
                line: 19,
                character: 2,
            },
        },
        children: Some(
            [
                DocumentSymbol {
                    name: "Test: bar",
                    detail: None,
                    kind: Function,
                    tags: None,
                    deprecated: None,
                    range: Range {
                        start: Position {
                            line: 17,
                            character: 0,
                        },
                        end: Position {
                            line: 19,
                            character: 2,
                        },
                    },
                    selection_range: Range {
                        start: Position {
                            line: 17,
                            character: 0,
                        },
                        end: Position {
                            line: 19,
                            character: 2,
                        },
                    },
                    children: Some(
                        [],
                    ),
                },
            ],
        ),
    },
]
